### PR TITLE
[Refactor] Error logging and refac of backend_factory

### PR DESCRIPTION
**Pull request by dominikandreasseitz**
1. `backend_factory` now only imports the requested backend instead of all backends as before
2. `config_factory` now only imports the config file instead of all backends and then getting the config from the requested one
3. Propagation of import errors

### CODE REVIEW

File: qadence/backends/api.py [hunk 1]
```diff
@@ -14,67 +21,47 @@ def backend_factory(
...
-    # Create the backend
-    backend_inst = BackendCls(
-        config=configuration
-        if configuration is not None
-        else BackendCls.default_configuration()  # type: ignore[attr-defined]
-    )
+        BackendCls = import_backend(backend)
+        default_config = BackendCls.default_configuration()
+        if configuration is None:
+            configuration = default_config
```
Comment (Roland-djee) [inline qadence/backends/api.py:28]:
```suggestion
            configuration = BackendCls.default_configuration()
```

File: qadence/backends/api.py [hunk 2]
```diff
@@ -14,67 +21,47 @@ def backend_factory(
...
+            if not isinstance(configuration, type(BackendCls.default_configuration())):
+                expected_cfg = BackendCls.default_configuration()
+                raise ValueError(
+                    f"Given config class '{type(configuration)}' does not match the backend",
+                    f" class: '{BackendCls}'. Expected: '{type(expected_cfg)}.'",
+                )
 
+        # Instantiate the backend
+        backend_inst = BackendCls(  # type: ignore[operator]
+            config=configuration
```
Comment (Roland-djee) [inline qadence/backends/api.py:42]: Is this check necessary ? It feels that at this point, we've exhausted all possibilities for `configuration` and excluded cases where it is undefined.

File: qadence/backends/api.py [hunk 3]
```diff
@@ -14,67 +21,47 @@ def backend_factory(
...
-    except KeyError:
-        raise NotImplementedError(f"The requested backend '{name}' is not implemented!")
-
-    BackendConfigCls = type(BackendCls.default_configuration())
-    return BackendConfigCls(**config)  # type: ignore[no-any-return]
+        BackendConfigCls = import_config(backend_name)
+        cfg = BackendConfigCls(**config)  # type: ignore[operator]
+    except Exception as e:
+        logger.debug(f"Unable to import config for backend {backend_name} due to {e}.")
+    return cfg
```
Comment (Roland-djee) [inline qadence/backends/api.py:67]:
```suggestion
    return BackendConfigCls(**config)
```

File: qadence/extensions.py [hunk 4]
```diff
@@ -2,41 +2,72 @@
...
-def _available_engines() -> dict:
+def import_config(backend_name: str | BackendName) -> BackendConfiguration:
+    module_path = f"qadence.backends.{backend_name}.config"
+    cfg: BackendConfiguration
+    try:
+        module = importlib.import_module(module_path)
+        cfg = getattr(module, "Configuration")
+    except (ModuleNotFoundError, ImportError) as e:
+        raise Exception(f"Failed to import backend config of {backend_name} due to {e}.")
+    return cfg
```
Comment (Roland-djee) [inline qadence/extensions.py:28]: If possible:

```suggestion
    return getattr(module, "Configuration")
```

File: qadence/extensions.py [hunk 5]
```diff
@@ -2,41 +2,72 @@
...
 backends_namespace = Template("qadence.backends.$name")
+BackendClsType = TypeVar("BackendClsType", bound=Backend)
+EngineClsType = TypeVar("EngineClsType", bound=DifferentiableBackend)
 
 logger = get_logger(__name__)
 
 
-def _available_engines() -> dict:
+def import_config(backend_name: str | BackendName) -> BackendConfiguration:
+    module_path = f"qadence.backends.{backend_name}.config"
```
Comment (Roland-djee) [inline qadence/extensions.py:21]: We should probably think of a better way to handle module paths than to hardcode them. Maybe we can use templated strings here ?

**Review by Roland-djee — APPROVED** [review]
Thanks @dominikandreasseitz. Looking good. Minor remarks.

**Comment by Roland-djee** [comment]
Hey @dominikandreasseitz can we merge this ?

**Comment by dominikandreasseitz** [comment]
> Hey @dominikandreasseitz can we merge this ?

@Roland-djee i think this would need to be adjusted to the new logging which isnt merged yet. i will have a look this week

**Comment by Roland-djee** [comment]
> > Hey @dominikandreasseitz can we merge this ?
> 
> @Roland-djee i think this would need to be adjusted to the new logging which isnt merged yet. i will have a look this week

Yeah, that was my impression too.

**Comment by dominikandreasseitz** [comment]
closing until #366 is merged

**Comment by Roland-djee** [comment]
Prompting you here @dominikandreasseitz as GH doesn't allow to request review from you for a reason.

**Comment by Roland-djee** [comment]
@dominikandreasseitz Would this be OK with you ?

**Comment by dominikandreasseitz** [comment]
> @dominikandreasseitz Would this be OK with you ?

@Roland-djee jep, we can merge imo

**Pull request merged**